### PR TITLE
Add Recipe: RAG with WatsonX and Langchain

### PR DESCRIPTION
**Pull request by fayvor**
[#121](https://github.com/ibm-granite-community/pm/issues/121) RAG Recipe using WatsonX, Chroma, Langchain:
- Download and chunk sample data
- Build Chroma DB from text embeddings
- Get WatsonX model with Langchain using IBM Provider adapter
- Build RAG chain with Langchain
- Issue sample query

### CODE REVIEW

**Comment by fayvor** [comment]
Someone please walk through this and confirm that you can reproduce using the WatsonX free tier.

File: recipes/Retrieval_Augmented_Generation/RAG_with_WatsonX_Chroma_Langchain.ipynb
Comment (deanwampler) [reply recipes/Retrieval_Augmented_Generation/RAG_with_WatsonX_Chroma_Langchain.ipynb:186]: As we discussed, maybe we can attempt to synthesize the URLs to click. It looks like https://cloud.ibm.com/iam/apikeys might work for API keys and https://dataplatform.cloud.ibm.com/wx/home?context=wx at least gets you to a page with a link to your Projects.

File: recipes/Retrieval_Augmented_Generation/RAG_with_WatsonX_Chroma_Langchain.ipynb
Comment (deanwampler) [reply recipes/Retrieval_Augmented_Generation/RAG_with_WatsonX_Chroma_Langchain.ipynb:159]: A warning like:
> **WARNING:** A string that starts with `cpd-apikey` or `APIKey` is _not_ the key value, but a name or id. When you _create the key_, you will need to save the value somewhere safe.

File: recipes/Retrieval_Augmented_Generation/RAG_with_WatsonX_Chroma_Langchain.ipynb
Comment (deanwampler) [inline recipes/Retrieval_Augmented_Generation/RAG_with_WatsonX_Chroma_Langchain.ipynb:784]: It took several minutes for an answer to come back! I don't know if it was something in my setup, but you might want to add a warning that sometimes it takes a long time to return an answer (in this "lightweight test setup"?).

**Review by deanwampler — APPROVED** [review]
Thanks for your help debugging my experience. We discussed some of the comments already... I'll approve the PR and let you do any additional fixes required. Let me know if you want a second review.

File: recipes/Retrieval_Augmented_Generation/RAG_with_WatsonX_Chroma_Langchain.ipynb
Comment (fayvor) [reply recipes/Retrieval_Augmented_Generation/RAG_with_WatsonX_Chroma_Langchain.ipynb:159]: I replaced `input()` with `getpass()` to hide the api key.

Comment (fayvor) [reply recipes/Retrieval_Augmented_Generation/RAG_with_WatsonX_Chroma_Langchain.ipynb:159]: Added the warning.

File: recipes/Retrieval_Augmented_Generation/RAG_with_WatsonX_Chroma_Langchain.ipynb
Comment (fayvor) [reply recipes/Retrieval_Augmented_Generation/RAG_with_WatsonX_Chroma_Langchain.ipynb:115]: Using Python 3.10, I was able to keep the versions of all libraries open.

File: recipes/Retrieval_Augmented_Generation/RAG_with_WatsonX_Chroma_Langchain.ipynb
Comment (fayvor) [reply recipes/Retrieval_Augmented_Generation/RAG_with_WatsonX_Chroma_Langchain.ipynb:186]: Here is the link to projects: https://dataplatform.cloud.ibm.com/projects/?context=wx

Comment (fayvor) [reply recipes/Retrieval_Augmented_Generation/RAG_with_WatsonX_Chroma_Langchain.ipynb:186]: I've added the urls to the places to get API key and Project Id.

File: recipes/Retrieval_Augmented_Generation/RAG_with_WatsonX_Chroma_Langchain.ipynb
Comment (fayvor) [reply recipes/Retrieval_Augmented_Generation/RAG_with_WatsonX_Chroma_Langchain.ipynb:784]: It was quick for me. Maybe longer for you since you are using the "Lite" plan? I'll put a note in.

Comment (fayvor) [reply recipes/Retrieval_Augmented_Generation/RAG_with_WatsonX_Chroma_Langchain.ipynb:784]: Does it come back faster the second time?

**Comment by deanwampler** [comment]
FYI,

I still see these warnings when pip installing:
```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
langchain-ibm 0.1.12 requires langchain-core<0.3,>=0.2.2, but you have langchain-core 0.3.1 which is incompatible.
langchain-community 0.2.15 requires langchain-core<0.3.0,>=0.2.37, but you have langchain-core 0.3.1 which is incompatible.
langchain-text-splitters 0.2.2 requires langchain-core<0.3.0,>=0.2.10, but you have langchain-core 0.3.1 which is incompatible.
langchain 0.2.16 requires langchain-core<0.3.0,>=0.2.38, but you have langchain-core 0.3.1 which is incompatible.
langchain-chroma 0.1.1 requires langchain-core<0.3,>=0.1.40, but you have langchain-core 0.3.1 which is incompatible.
Successfully installed langchain-core-0.3.1 langsmith-0.1.121
Successfully installed langchain-core-0.2.40
Requirement already satisfied: six>=1.10.0 in /opt/homebrew/Caskroom/miniforge/base/envs/granite-code/lib/python3.11/site-packages (from lomond->ibm-watsonx-ai) (1.16.0)
Requirement already satisfied: six>=1.10.0 in /opt/homebrew/Caskroom/miniforge/base/envs/granite-code/lib/python3.11/site-packages (from lomond->ibm-watsonx-ai<2.0.0,>=1.0.8->langchain_ibm) (1.16.0)
Requirement already satisfied: wget in /opt/homebrew/Caskroom/miniforge/base/envs/granite-code/lib/python3.11/site-packages (3.2)
Requirement already satisfied: mpmath>=0.19 in /opt/homebrew/Caskroom/miniforge/base/envs/granite-code/lib/python3.11/site-packages (from sympy->torch>=1.11.0->sentence-transformers) (1.3.0)
Requirement already satisfied: pyasn1<0.7.0,>=0.4.6 in /opt/homebrew/Caskroom/miniforge/base/envs/granite-code/lib/python3.11/site-packages (from pyasn1-modules>=0.2.1->google-auth>=1.0.1->kubernetes>=28.1.0->chromadb) (0.6.0)
Requirement already satisfied: typing-extensions>=4.6.1 in /opt/homebrew/Caskroom/miniforge/base/envs/granite-code/lib/python3.11/site-packages (from pydantic) (4.11.0)
Requirement already satisfied: typing-extensions>=4.6.0 in /opt/homebrew/Caskroom/miniforge/base/envs/granite-code/lib/python3.11/site-packages (from sqlalchemy) (4.11.0)
Requirement already satisfied: mypy-extensions>=0.3.0 in /opt/homebrew/Caskroom/miniforge/base/envs/granite-code/lib/python3.11/site-packages (from typing-inspect<1,>=0.4.0->dataclasses-json<0.7,>=0.5.7->langchain-community) (1.0.0)
```

However, everything still appeared to work successfully. You might consider adding a note in the comments above that some `ERROR`s might be reported, but they should not cause issues. (Does anyone else see these messages? I wonder if I see them because I'm using a condo environment?)

The `WATSONX_PROJECT_ID` is still echoed in the open. This is probably not harmful, but still might worry the user.

Some cells still take minutes. I like the note before the last cell that it can take minutes. How about adding the same warning before the following cell, too?

```python
from langchain_ibm import WatsonxEmbeddings
from ibm_watsonx_ai.foundation_models.utils.enums import EmbeddingTypes

embeddings = WatsonxEmbeddings(
    model_id=EmbeddingTypes.IBM_SLATE_30M_ENG.value,
    url=credentials["url"],
    apikey=credentials["apikey"],
    project_id=project_id
    )
docsearch = Chroma.from_documents(texts, embeddings)
```

**Comment by fayvor** [comment]
> I still see these warnings when pip installing:

@deanwampler what version of python are you using? I tried this with a fresh virtualenv of `3.10.15`.

**Comment by deanwampler** [comment]
> > I still see these warnings when pip installing:
> 
> @deanwampler what version of python are you using? I tried this with a fresh virtualenv of `3.10.15`.

3.11.9. I now think it was something about my environment. I tried with two clean conda environments, one using 3.10 and the other using 3.11, and both got through the pip installs with no warnings and ran to completion. Interesting that the 3.11 environment returned more quickly for the last cell than before. The 3.10 environment might have as well, but I wasn't watching closely.

So, what I suggested earlier about adding a warning about potential errors, that they are probably "okay", could perhaps be changed to this: _if you get errors during the pip installations, try creating a clean environment with `virtualenv` or Anaconda._

**Comment by rawkintrevo** [comment]
recipes/Retrieval_Augmented_Generation/RAG_with_WatsonX_Chroma_Langchain.ipynb

**Review by rawkintrevo — CHANGES_REQUESTED** [review]
please rebase on main and add `recipes/Retrieval_Augmented_Generation/RAG_with_WatsonX_Chroma_Langchain.ipynb`
to the appropriate file in `.git/workflows`

**Pull request closed without merging**
